server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +412,300 @@ public int hashCode() {
+         *
+         * It increments the staleKeysCount by the count of the CleanupKey in the cleanupKeyToCountMap.
+         * If the CleanupKey's readerCacheKeyId is null or the CleanupKey's entity is not open, it increments the staleKeysCount
+         * by the total count of keys associated with the CleanupKey's ShardId in the cleanupKeyToCountMap and removes the ShardId from the map.
+         *
+         * @param cleanupKey the CleanupKey that has been marked for cleanup
+         */
+        private void updateStaleKeysCount(CleanupKey cleanupKey) {
+            if (stalenessThreshold == 0.0) {
+                return;
+            }
+            IndexShard indexShard = (IndexShard) cleanupKey.entity.getCacheIdentity();
+            if (indexShard == null) {
+                logger.warn("IndexShard is null for CleanupKey: {}", cleanupKey.readerCacheKeyId);
+                return;
+            }
+            ShardId shardId = indexShard.shardId();
+
+            // Using computeIfPresent to atomically operate on the countMap for a given shardId
+            cleanupKeyToCountMap.computeIfPresent(shardId, (key, countMap) -> {
+                if (cleanupKey.readerCacheKeyId == null) {
+                    // Aggregate and add to staleKeysCount atomically if readerCacheKeyId is null
+                    int totalSum = countMap.values().stream().mapToInt(Integer::intValue).sum();
+                    staleKeysCount.addAndGet(totalSum);
+                    // Return null to automatically remove the mapping for shardId
+                    return null;
+                } else {
+                    // Update staleKeysCount based on specific readerCacheKeyId, then remove it from the countMap
+                    countMap.computeIfPresent(cleanupKey.readerCacheKeyId, (k, v) -> {
+                        staleKeysCount.addAndGet(v);
+                        // Return null to remove the key after updating staleKeysCount
+                        return null;
+                    });
+
+                    // Check if countMap is empty after removal to decide if we need to remove the shardId entry
+                    if (countMap.isEmpty()) {
+                        return null; // Returning null removes the entry for shardId
+                    }
+                }
+                return countMap; // Return the modified countMap to keep the mapping
+            });
+        }
+
+        AtomicInteger getStaleKeysCountForTesting() {
+            return staleKeysCount;
         }
-        if (!currentKeysToClean.isEmpty() || !currentFullClean.isEmpty()) {
+
+        /**
+         * Clean cache based on stalenessThreshold
+         */
+        void cleanCache() {
+            cleanCache(stalenessThreshold);
+        }
+
+        /**
+         * Force Clean cache without checking stalenessThreshold
+         */
+        private void forceCleanCache() {
+            cleanCache(0);
```

Comment:
Why not consider staleness here as well?
If we don't consider staleness, then what's the point of updating staleKeysCount on clean(CacheEntity)?

Reply:
forceCleanCache is called from the clear cache API, 
Earlier we used to 
```
    void clear(CacheEntity entity) {
        keysToClean.add(new CleanupKey(entity, null));
        cleanCache();
    }
```
Since this is a clear intent from the user to clean cache, we force clean the cache without deciding on staleness factor. 
If we also do this based on threshold, there is no way for the users to clear cache on choice.